wordpress-mobile/WordPress-Android
# List all likes in notifications

Part of the #14137 (companion Fluxc PR [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1935))

This PR shows all who liked a post or comment in a notification.

Hi @mattmiklic 👋 . The user profile bottom sheet is being added in another PR but happy if you could have a quick sanity check on the images below, thanks 🙇 ! 
A few thins to highlight:

1. I kept the header element directly in the list (so it moves up and down with the elements and does not stay sticky on top). I did this to keep the list consistent with what we have right now in the other lists of notifications details (like followers). 
2. There is an extra margin in the first liker element that is introduced by the header element. I kept it to be consistent with the other kind of lists for now.
3. About the followers list; currently the two lists (likers and followers) are rendered in a different way, probably we will need to align the followers list to this likers list, but I guess it's ok to release the likers feature first and align the followers later? wdyt?

| Description | Image |
|---|---|
| Dedicated fragment for Engaged People (to be reused for Followers as well in future) | <p align=center><image src="https://user-images.githubusercontent.com/47797566/112737547-a0f2e180-8f5b-11eb-8f6d-30b323420431.png" width=300><p> |
| A loading state has been introduced | <p align=center><image src="https://user-images.githubusercontent.com/47797566/112737664-6178c500-8f5c-11eb-8807-33b0b85e000e.png" width=300><p> |
| Empty state for no network | <p align=center><image src="https://user-images.githubusercontent.com/47797566/112737614-12cb2b00-8f5c-11eb-8fd3-6b37bd27aa36.png" width=300><p> |
| Empty state when errors in getting likes | <p align=center><image src="https://user-images.githubusercontent.com/47797566/112810621-7f891700-907b-11eb-814b-09f108bdeb79.png" width=300><p> |

**REVIEW NOTES**
1. User profile bottom sheet, unit tests and analytics will be added in dedicated PR(s)
2. Pagination will be aligned with the server side pagination of the endpoint when it is available; currently likes are taken from the endpoint up to the 90 limit that is coded in the endpoint sources
3. The order of likers in the list is currently different than what we get from notifications itself. An endpoint modification is ongoing to align this (cc @aerych).
4. About the confetti effect that is present in the `NotificationsDetailsListFragment`: in my understanding confetti celebration effect does not apply to `NOTE_LIKE_TYPE` and `NOTE_COMMENT_LIKE_TYPE` but only to `NOTE_VIEW_MILESTONE` so we should be fine to not include logic for that in these likers list (cc @planarvoid for a sanity check 🙇 )

## To test
The feature is behind a Feature Flag to which we are adding in next PRs. Good to compare the test below side by side with a develop installation.

- Open a notification from the ALL notification filter and check the notifications is loaded correctly
- Try to swipe in between the opened notification details and check the UI is rendered correctly in both expected notifications order and content.
- When you are swiping to a `LIKE` notification (post or comment) check you get the loading state and you get the likers list. Possibly check also list with lot of post or comment likers
- Do same checks above opening from the `LIKES` notification filter
- Smoke test trying to open notifications details from different type of notifications
- Check the empty states illustrated in the images in the beginning of the PR description
- From a post like details (and later also from a comment like details) check side by side that you get the same opened screens tapping on the 
    - header author avatar
    - header itself
    - an liker item in the list (for this point and tapping on the header elements would be good to force (by debugger or finding a suitable notification) to get an webview renderer the site or an app reader screen
- With feature flag off check that we are having same behaviour as before. Some things to check: notifications list 

PR submission checklist:
